backends/fs/kolibrios/kolibrios-fs.cpp

```diff
@@ -0,0 +1,310 @@
+	uint32 version;
+	uint32 number_of_placed_blocks;
+	uint32 total_number_of_blocks;
+	byte padding[20];
+};
+
+struct kol_readdir_result_bdfe {
+	uint32 attributes;
+	uint32 encoding;
+	uint64 ctime;
+	uint64 atime;
+	uint64 mtime;
+	uint64 size;
+};
+
+struct kol_readdir_result_entry {
+	kol_readdir_result_bdfe bdfe;
+	char fileName[520];
+};
+
+struct kol_readdir_result
+{
+	kol_readdir_result_header header;
+	kol_readdir_result_entry entries[0];
+};
+struct kol_readdir_request
+{
+	uint32 subfunc;
+	uint32 starting_block;
+	uint32 encoding;
+	uint32 blocks_to_read;
+	struct kol_readdir_result *result;
+	byte   p20;
+	const char *path;
+};
+#pragma pack(pop)
+
+namespace {
+
+// opendir/readdir are buggy. And they encourage using syscalls. Whatever
+int kol_file_70(kol_readdir_request *k)
+{
+	int ret, num_read;
+	asm volatile ("int $0x40": "=a"(ret), "=b"(num_read) :"a"(70), "b"(k));
+	return ret;
+}
+
+bool getFileAttrs(Common::String path, uint32& attrs) {
+	fileinfo_t info;
+
+	memset(&info, 0, sizeof(info));
+	info.attr = 0x10;
+	if(get_fileinfo(path.c_str(), &info)) {
+		attrs = 0;
+		return false;
+	}
+
+	attrs = info.attr;
+	return true;
+}
+
+}
+
+bool KolibriOSFilesystemNode::exists() const {
+	uint32 attrs;
+	return getFileAttrs(_path, attrs);
+}
+
+bool KolibriOSFilesystemNode::isReadable() const {
+	uint32 attrs;
+	return getFileAttrs(_path, attrs);
+}
+
+bool KolibriOSFilesystemNode::isWritable() const {
+	uint32 attrs;
+	return getFileAttrs(_path, attrs) && !(attrs & 0x01);
+}
+
+void KolibriOSFilesystemNode::setFlags() {
+	_isValid = getFileAttrs(_path, _attributes);
+}
+
+KolibriOSFilesystemNode::KolibriOSFilesystemNode(const Common::String &p) {
+	assert(p.size() > 0);
+
+	_path = p;
+
+	// Normalize the path (that is, remove unneeded slashes etc.)
+	_path = Common::normalizePath(_path, '/');
+	_displayName = Common::lastPathComponent(_path, '/');
+
+	setFlags();
+}
+
+AbstractFSNode *KolibriOSFilesystemNode::getChild(const Common::String &n) const {
+	assert(!_path.empty());
+	assert(isDirectory());
+
+	// Make sure the string contains no slashes
+	assert(!n.contains('/'));
+
+	// We assume here that _path is already normalized (hence don't bother to call
+	//  Common::normalizePath on the final path).
+	Common::String newPath(_path);
+	if (_path.lastChar() != '/')
+		newPath += '/';
+	newPath += n;
+
+	return makeNode(newPath);
+}
+
+bool KolibriOSFilesystemNode::getChildren(AbstractFSList &myList, ListMode mode, bool hidden) const {
+	kol_readdir_request k70;
+	int result; 
+
+	k70.subfunc = 1;
+	k70.starting_block = 0;
+	k70.encoding = 3;
+	k70.blocks_to_read = 2;  // just for test exist & read number of entries
+	k70.result = (kol_readdir_result *) malloc(sizeof(kol_readdir_result_header)+k70.blocks_to_read*sizeof(kol_readdir_result_entry));
+	k70.p20 = 0;
+
+	k70.path = _path.c_str();
+
+	result = kol_file_70(&k70);
+	if ( !((result==0) || (result==6)) ) // check does the directory exists
+	{
```

Comment:
This whole file has to be reformatted. It looks like borrowed from somewhere. Was the license compatible?

Reply:
I thought it was BSDL but turns out no license was specified. As ther code was simple and not very good to begin with, I rewrote it. Structures were rewritten from documentation already because original used numbered fields which are completely unreadable